src/main/java/com/cleanroommc/kirino/engine/render/task/data/KDTree.java

```diff
@@ -0,0 +1,209 @@
+package com.cleanroommc.kirino.engine.render.task.data;
+
+import com.cleanroommc.kirino.engine.render.geometry.AABB;
+import com.cleanroommc.kirino.engine.render.geometry.Block;
+import com.cleanroommc.kirino.utils.QuantileUtils;
+import com.google.common.base.Preconditions;
+import it.unimi.dsi.fastutil.objects.ReferenceLinkedOpenHashSet;
+import kotlin.random.Random;
+import org.joml.Vector3i;
+import org.jspecify.annotations.NonNull;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Set;
+
+public class KDTree {
+
+    private static final int MAX_BLOCKS_IN_LEAF = 32;
+    private static final int DIMENSIONS = 3;
+
+    private Node root = null;
+    private final Set<Node> leaves = new ReferenceLinkedOpenHashSet<>(); // Linked since there will be a lot of deletions
```

Comment:
```suggestion
    private final Set<Node> leaves = new ReferenceLinkedOpenHashSet<>(); // linked since there will be a lot of deletions
```
<br>
<img width="1454" height="1342" alt="image" src="https://github.com/user-attachments/assets/a883563c-c1d1-474d-b10f-7c8d5137c333" />
<br>
sorry to be a bit picky here but it's necessary, and pls fix other comment-line formatting and single-space-after-comma formatting